loopy/transform/reduction.py

```diff
@@ -0,0 +1,292 @@
+                                                  List[Tpart]]:
+    """
+    Use a predicate to partition entries into false entries and true
+    entries
+    """
+    # Inspired from https://docs.python.org/3/library/itertools.html
+    # partition(is_odd, range(10)) --> 0 2 4 6 8   and  1 3 5 7 9
+    from itertools import tee, filterfalse
+    t1, t2 = tee(iterable)
+    return list(filterfalse(pred, t1)), list(filter(pred, t2))
+
+# }}}
+
+
+# {{{ hoist_reduction_invariant_terms
+
+class EinsumTermsHoister(IdentityMapper):
+    """
+    Mapper to hoist products out of a sum-reduction.
+
+    .. attribute:: reduction_inames
+
+        Inames of the reduction expressions to perform the hoisting.
+    """
+    def __init__(self, reduction_inames: FrozenSet[str]):
+        super().__init__()
+        self.reduction_inames = reduction_inames
+
+    # type-ignore-reason: super-class.map_reduction returns 'Any'
+    def map_reduction(self, expr: Reduction  # type: ignore[override]
+                      ) -> p.Expression:
+        if frozenset(expr.inames) != self.reduction_inames:
+            return super().map_reduction(expr)
+
+        from loopy.library.reduction import SumReductionOperation
+        from loopy.symbolic import get_dependencies
+        if isinstance(expr.operation, SumReductionOperation):
+            if isinstance(expr.expr, p.Product):
+                from pymbolic.primitives import flattened_product
+                multiplicative_terms = (flattened_product(self.rec(expr.expr)
+                                                          .children)
+                                        .children)
+            else:
+                multiplicative_terms = (expr.expr,)
+
+            invariants, variants = partition(lambda x: (get_dependencies(x)
+                                                        & self.reduction_inames),
+                                             multiplicative_terms)
+            if not variants:
+                # -> everything is invariant
+                return self.rec(expr.expr) * Reduction(
+                    expr.operation,
+                    inames=expr.inames,
+                    expr=1,  # FIXME: invalid dtype (not sure how?)
+                    allow_simultaneous=expr.allow_simultaneous)
+            if not invariants:
+                # -> nothing to hoist
+                return Reduction(
+                    expr.operation,
+                    inames=expr.inames,
+                    expr=self.rec(expr.expr),
+                    allow_simultaneous=expr.allow_simultaneous)
+
+            return p.Product(tuple(invariants)) * Reduction(
+                expr.operation,
+                inames=expr.inames,
+                expr=p.Product(tuple(variants)),
+                allow_simultaneous=expr.allow_simultaneous)
+        else:
+            return super().map_reduction(expr)
+
+
+def hoist_invariant_multiplicative_terms_in_sum_reduction(
```

Comment:
What about other operations that satisfy a distributive property? (Specify classes maybe? We don't need to enumerate things that obey the distributive law. We can't, not comprehensively, anyway.)